### PR TITLE
feat: Add support to prepare CoreDNS for Helm management, enable cluster proportional autoscaler for CoreDNS

### LINKED ISSUE #891: Fargate wildcard profile support
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

Modifying existing examples to show fargate feature using wildcard support for profile selection: https://aws.amazon.com/about-aws/whats-new/2022/08/wildcard-support-amazon-eks-fargate-profile-selectors/

### LINKED ISSUE #1028: Opinionated defaults for CoreDNS
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

- Better support for CoreDNS with less configuration required by users

#### Describe the solution you would like

- Enable to toggle CoreDNS support on Fargate only clusters 
- Enable cluster proportional autoscaler on CoreDNS by default with ability for users to use their own CPA

### PR DESCRIPTION
**Pull request by bryantbiggs**
### What does this PR do?

- Add support to prep CoreDNS for Helm management within the CoreDNS addon. This improves the usability experience when deploying self-managed CoreDNS via the addon
- Add support for CoreDNS cluster proportional autoscaler. This is now enabled by default with some sane default settings when using either EKS managed CoreDNS  or self-managed via the addon
- Add `time_sleep` resource to CoreDNS addon to ensure proper sequencing of events within the addon (all of the following are subject to being enabled, but regardless the order still stands now):
	1. Remove the default deployment of CoreDNS
	2. Deploy the self-managed (Helm addon) CoreDNS deployment
	3. Enable cluster proportional autoscaler
- Add new variable `data_plane_wait_arn` which can be used to coordinate the dependency ordering between the EKS cluster creation and addon creation. By passing in an ARN from a node group or Fargate profile (or an array of ARNs using `join()`, the `time_sleep` resource will wait for the ARN(s) passed before starting a brief sleep duration and then allowing the deployment of the addons. This makes it possible to do a single `terraform apply` with the proper ordering and at a later date we can pass additional information through `time_sleep` to ensure proper ordering when performing cluster upgrades without the need for additional information from users (*I believe - TBD).
- Add support for the 2048 sample/demo app as an "addon" to be able to re-use in other examples. Removing the use of the kubectl provider for this and only using the Kubernetes provider reduces the number of providers needed in the examples for this demo app as well
- Update Fargate example blueprint
	- Enable 2048 demo app via addon
	- Update CoreDNS to be self-managed to run on Fargate
	- Show example of Fargate wildcard profile usage for the demo app
	- Enable `fargate-fluentbit` addon for Fargate logging (removes warnings from pods + we should have logging)
	
### Motivation

- Improve user experience with CoreDNS as well as Fargate based usage
- Resolves #1028 
- Resolves #891 

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

### CODE REVIEW

File: examples/fargate-serverless/main.tf [hunk 1]
```diff
@@ -82,11 +73,11 @@ module "eks_blueprints" {
...
     }
     # Sample application
-    alb_sample_app = {
-      fargate_profile_name = "alb-sample-app"
+    app = {
+      fargate_profile_name = "app-wildcard"
       fargate_profile_namespaces = [
         {
-          namespace = local.sample_app_namespace
+          namespace = "app-*"
```
Comment (bryantbiggs) [inline examples/fargate-serverless/main.tf:80]: Resolves https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/891

File: examples/fargate-serverless/main.tf [hunk 2]
```diff
@@ -103,6 +94,9 @@ module "eks_blueprints_kubernetes_addons" {
   eks_oidc_provider    = module.eks_blueprints.oidc_provider
   eks_cluster_version  = module.eks_blueprints.eks_cluster_version
 
+  # Wait on the `kube-system` profile before provisioning addons
+  data_plane_wait_arn = module.eks_blueprints.fargate_profiles["kube_system"].eks_fargate_profile_arn
```
Comment (bryantbiggs) [inline examples/fargate-serverless/main.tf:98]: pretty nifty - addons don't start provisioning until the profile has provisioned

File: examples/fargate-serverless/main.tf [hunk 3]
```diff
@@ -152,82 +149,10 @@ data "aws_eks_addon_version" "latest" {
   most_recent        = true
 }
 
-#---------------------------------------------------------------
-# Modifying CoreDNS for Fargate
-#---------------------------------------------------------------
-
-locals {
```
Comment (bryantbiggs) [inline examples/fargate-serverless/main.tf:159]: All this lives in the CoreDNS addon now so users do not need to re-create if they want to use this method of assuming control of CoreDNS with Helm

File: examples/fargate-serverless/main.tf [hunk 4]
```diff
@@ -252,31 +177,12 @@ module "vpc" {
   default_security_group_tags   = { Name = "${local.name}-default" }
 
   public_subnet_tags = {
-    "kubernetes.io/cluster/${local.name}" = "shared"
```
Comment (bryantbiggs) [inline examples/fargate-serverless/main.tf:255]: no longer required with recent versions of ALB controller

File: modules/kubernetes-addons/cluster-proportional-autoscaler/variables.tf [hunk 5]
```diff
@@ -22,7 +22,5 @@ variable "addon_context" {
     eks_oidc_issuer_url            = string
     eks_oidc_provider_arn          = string
     tags                           = map(string)
-    irsa_iam_role_path             = string
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/cluster-proportional-autoscaler/variables.tf:25]: not used

File: modules/kubernetes-addons/aws-coredns/main.tf [hunk 6]
```diff
@@ -60,9 +60,144 @@ module "helm_addon" {
     {
       name  = "serviceAccount.create"
       value = true
+    },
+    {
+      # This simply maps a dependency for Terraform to ensure the default deployment is
+      # removed first (if enabled) before provisioning self-managed version (if enabled)
+      name  = "blueprints.connection"
+      value = try(null_resource.remove_default_coredns_deployment[0].id, "none")
```
Comment (askulkarni2) [inline modules/kubernetes-addons/aws-coredns/main.tf:68]: nice!

**Review by askulkarni2 — COMMENTED** [review]
One question, other wise the PR is rock solid.

File: examples/fargate-serverless/main.tf (hunk 4 above)
Comment (FernandoMiguel) [reply examples/fargate-serverless/main.tf:255]: interesting. where is this documented? 
we are on the latest release of AWS LB Controller, and still tag out SGs

File: examples/fargate-serverless/main.tf (hunk 2 above)
Comment (FernandoMiguel) [reply examples/fargate-serverless/main.tf:98]: do you find this approach better than the use of `depends_on` ?
I assume it replaces it, corrrect?

File: examples/fargate-serverless/main.tf (hunk 4 above)
Comment (bryantbiggs) [reply examples/fargate-serverless/main.tf:255]: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/docs/deploy/subnet_discovery.md#common-tag

Comment (bryantbiggs) [reply examples/fargate-serverless/main.tf:255]: Security group tags is separate, and that is still required. This is just subnet tags

File: modules/kubernetes-addons/variables.tf [hunk 7]
```diff
@@ -15,6 +15,12 @@ variable "eks_worker_security_group_id" {
   default     = ""
 }
 
+variable "data_plane_wait_arn" {
+  description = "Addon deployment will not proceed until this value is known. Set to node group/Fargate profile ARN to wait for data plane to be ready before provisioning addons"
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply modules/kubernetes-addons/variables.tf:19]: A list can still be used if you wrap it in a `join()` function. Ideally, you wouldn't want to wait for all of the node groups and Fargate profiles to come up, you just need one where addons can start deploying. Also, using one over using many will keep the potential for updates to a minimum (the more that are provided in the trigger, the more times there could be a re-trigger when its not necessary)

File: examples/fargate-serverless/main.tf (hunk 4 above)
Comment (FernandoMiguel) [reply examples/fargate-serverless/main.tf:255]: right right right!

**Review by askulkarni2 — APPROVED** [review]
LGTM!

**Pull request merged**